include/seqan/seq_io/bam_sam.h

```diff
@@ -0,0 +1,357 @@
+
+    remainingBytes -= sizeof(BamAlignmentRecordCore) + recordCore._l_qname +
+                      recordCore._n_cigar * 4 + (recordCore._l_qseq + 1) / 2 + recordCore._l_qseq;
+    SEQAN_ASSERT_GEQ(remainingBytes, 0);
+
+    // query name.
+    resize(meta, recordCore._l_qname - 1, Exact());
+    arrayCopyForward(it, it + recordCore._l_qname - 1, begin(meta, Standard()));
+    it += recordCore._l_qname;
+
+    if (prevQName == meta)
+        return 0;
+
+    // skip cigar string.
+    it += sizeof(uint32_t) * recordCore._n_cigar;
+
+    // query sequence.
+    resize(seq, recordCore._l_qseq, Exact());
+    TSeqIter sit = begin(seq, Standard());
+    TSeqIter sitEnd = sit + (recordCore._l_qseq & ~1);
+    while (sit != sitEnd)
+    {
+        unsigned char ui = getValue(it);
+        ++it;
+        assignValue(sit, Iupac(ui >> 4));
+        ++sit;
+        assignValue(sit, Iupac(ui & 0x0f));
+        ++sit;
+    }
+    if (recordCore._l_qseq & 1)
+        *sit++ = Iupac((uint8_t)*it++ >> 4);
+    return recordCore._l_qseq; // returns length of query sequence. Needed for reading quality
+}
+
+
+// ----------------------------------------------------------------------------
+// Function readRecord
+// ----------------------------------------------------------------------------
+
+template <typename TIdString, typename TSeqString,
+          typename TForwardIter>
+inline bool
+_readRecord(TIdString & meta,
+            TSeqString & seq,
+            TIdString & prevQName,
+            TForwardIter & iter,
+            Sam const & /*tag*/)
+{
+    typedef typename Value<TSeqString>::Type                                TSeqAlphabet;
+    typedef typename SamIgnoreOrAssertFunctor_<TSeqAlphabet>::Type          TSeqIgnoreOrAssert;
+
+    // fail, if we read "@" (did you miss to call readRecord(header, bamFile) first?)
+    if (nextIs(iter, SamHeader()))
+        SEQAN_THROW(ParseError("Unexpected SAM header encountered."));
+
+    OrFunctor<IsTab, AssertFunctor<NotFunctor<IsNewline>, ParseError, Sam> > nextEntry;
+
+    clear(meta);
+    clear(seq);
+
+    // QNAME
+    readUntil(meta, iter, nextEntry);
+    skipOne(iter, IsTab());
+
+    // if we have seen this sequence before, return
+    if (meta == prevQName)
+        return false;
+
+    // Skip FLAG, RNAME, POS, MAPQ, CIGAR, RNEXT, PNEXT, TLEN
+    for (unsigned i = 0; i < 8; ++i)
+    {
+        skipUntil(iter, IsTab());
+        skipOne(iter, IsTab());
+    }
+
+    // SEQ
+    readUntil(seq, iter, nextEntry, TSeqIgnoreOrAssert());
+    // Handle case of missing sequence:  Clear seq string as documented.
+    if (seq == "*")
+        clear(seq);
+    skipOne(iter, IsTab());
+    return true;
+}
+
+template <typename TSpec, typename TIdString, typename TSeqString, typename TQualString>
+inline void readRecord(TIdString & meta, TSeqString & seq, TQualString & qual, FormattedFile<Fastq, Input, TSpec>&file, Sam)
+{
+    if (!(context(file).hasReadHeader))
+    {
+        skipHeader(file);
```

Comment:
call type-specific function here, see above
also wrap the if clause in `SEQAN_UNLIKELY(..)` since it will always only be true one time.